### PR TITLE
MRG: Tweak reject API

### LINKED ISSUE #1452: ENH/FIX: allow for reject API  that is more consistent with MNE C
In MNE-C it is possible to specify epoching parameters using a config file like this:
```
average {
outfile __OUTFILE__
eventfile __EVEFILE__
logfile __LOGFILE__
grdReject 200e-12
magReject 4e-12
eegReject -1
eogReject -1
emgReject -1
name "WordOnsets"

category {
name "VerbNP"
event 71 
tmin -0.3
tmax 1.2
}
}
```

MNE-Python new comers might be tempted to construct a reject dict like this:

```Python

reject = dict(mag=4e-12, grad=200e-12, eog=-1, ecg=-1)

# or

reject = dict(mag=4e-12, grad=200e-12, eog=None, ecg=None)
```

Which can / will lead to unexpected results, e.g. in Python comparisons None is always smaller than any numerical value ...

I would suggest to either allow for the MNE C style API or throw nice + concise errors once values like -1 or None are passed.

cc @agramfort @Eric89GXL 

Thanks to @bmaess to bringing this to my awareness

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #1452.

### CODE REVIEW

**Comment by Eric89GXL** [comment]
Ready for review/merge from my end. I think it's what was settled on for #1452.

**Comment by jasmainak** [comment]
What exactly are you proposing? Do you want to allow more keys than channel types present in the data?

**Comment by Eric89GXL** [comment]
Yes, see linked issue

**Comment by Eric89GXL** [comment]
But in short:

1. Throw error for `None` or `val < 0` for reject values to ensure people give proper values (can use `np.inf` or omit for "ignore")

2. Allow users to say `allow_missing_reject_keys=True` to allow them to pass a dict with channel type keys that do not exist in data.

**Comment by jasmainak** [comment]
We have been allowing `None` for reject values when we don't reject, no? Doesn't this constitute an API change and requires deprecation?

I'm not sure about 2. I think @agramfort had some argument against it because he wanted users to be aware of what they were doing, but having this argument would address that concern if it was always `False` by default.

File: mne/epochs.py [hunk 1]
```diff
@@ -444,49 +445,56 @@ def apply_baseline(self, baseline, verbose=None):
...
-            for key in set(new_flat.keys()).union(old_flat.keys()):
-                old = old_flat.get(key, -np.inf)
-                new = new_flat.get(key, -np.inf)
-                if new < old:
-                    raise ValueError(bad_msg.format(kind='flat', key=key,
-                                                    new=new, old=old, op='<'))
+            for rf in (reject, flat):
+                if key in rf:
+                    if len(idx[key]) == 0:
+                        if not self.allow_missing_reject_keys:
```
Comment (jasmainak) [inline mne/epochs.py]: since you have no `else` statements, couldn't you combine these nested `if` statements into one?

File: mne/epochs.py [hunk 2]
```diff
@@ -1790,14 +1801,14 @@ def __init__(self, raw, events, event_id, tmin, tmax, baseline=(None, 0),
...
         # call _BaseEpochs constructor
-        super(Epochs, self).__init__(info, None, events, event_id, tmin, tmax,
-                                     baseline=baseline, raw=raw, picks=picks,
-                                     name=name, reject=reject, flat=flat,
-                                     decim=decim, reject_tmin=reject_tmin,
-                                     reject_tmax=reject_tmax, detrend=detrend,
-                                     add_eeg_ref=add_eeg_ref, proj=proj,
-                                     on_missing=on_missing,
-                                     preload_at_end=preload, verbose=verbose)
+        super(Epochs, self).__init__(
```
Comment (jasmainak) [inline mne/epochs.py]: why did you change these lines? I don't see a difference except at the last line ...

File: mne/tests/test_epochs.py [hunk 3]
```diff
@@ -76,6 +76,14 @@ def test_reject():
                   picks=picks, preload=False, reject='foo')
     assert_raises(ValueError, Epochs, raw, events, event_id, tmin, tmax,
                   picks=picks_meg, preload=False, reject=dict(eeg=1.))
+    with warnings.catch_warnings(record=True) as w:
+        Epochs(raw, events, event_id, tmin, tmax, picks=picks_meg,
+               preload=False, reject=dict(eeg=1.),
+               allow_missing_reject_keys=True)
```
Comment (jasmainak) [inline mne/tests/test_epochs.py]: for compatibility, I suppose we should also add this to `drop_bad_epochs`? It would be weird to have it only in the `Epochs` constructor ...

**Comment by jasmainak** [comment]
@Eric89GXL : I fear disallowing `None` is too drastic. It will crash all my (and I fear for many other users too) scripts ...

**Comment by jasmainak** [comment]
My bad, I just noticed where you disallowed `None`. It looks good there ...

File: mne/tests/test_epochs.py (hunk 3 above)
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: It's a property of the class so users can change it themselves before
calling the function of they want. The only reason to put it in the
constructor is because the reject properties can also be set during
construction.

File: mne/epochs.py (hunk 2 above)
Comment (Eric89GXL) [inline mne/epochs.py]: More compact representation of the parameters

**Comment by Eric89GXL** [comment]
It's False by default so I think it should be safe enough.

**Comment by agramfort** [comment]
hum not super convinced

is this really needed?

it is also yet another parameter in Epochs...

**Comment by Eric89GXL** [comment]
We should at least keep the None/-1 changes since those are safeguards. I
don't really care about the "missing" stuff but thought that was the
consensus from the issue. I'll remove it though since it's easy to add back
later.

**Comment by agramfort** [comment]
so now we would be able to pass eeg=None instead of not passing it?

**Comment by Eric89GXL** [comment]
No, that will throw an error. So will `eeg=-1`.

**Comment by Eric89GXL** [comment]
Basically this adds safeguards against those cases.

**Comment by Eric89GXL** [comment]
Changes pushed to eliminate `allow_missing_reject_keys`, now it's just a safeguarding PR.

File: mne/tests/test_epochs.py [hunk 4]
```diff
@@ -76,6 +76,9 @@ def test_reject():
                   picks=picks, preload=False, reject='foo')
     assert_raises(ValueError, Epochs, raw, events, event_id, tmin, tmax,
                   picks=picks_meg, preload=False, reject=dict(eeg=1.))
+    for val in (None, -1):  # protect against older MNE-C types
+        assert_raises(ValueError, Epochs, raw, events, event_id, tmin, tmax,
+                      picks=picks_meg, preload=False, reject=dict(grad=val))
```
Comment (Eric89GXL) [inline mne/tests/test_epochs.py]: @agramfort see here for use

Comment (agramfort) [inline mne/tests/test_epochs.py]: same applies for flat right?

File: doc/whats_new.rst [hunk 5]
```diff
@@ -37,6 +37,8 @@ BUG
 
     - Fix bug that prevented homogeneous bem surfaces to be displayed in HTML reports by `Denis Engemann`_
 
+    - Added safeguards against ``None`` and negative values in reject parameters in :class:`mne.Epochs` by `Eric Larson`_
```
Comment (agramfort) [inline doc/whats_new.rst]: same applies for flat right?

**Comment by agramfort** [comment]
besides LGTM

File: mne/tests/test_epochs.py (hunk 4 above)
Comment (jasmainak) [inline mne/tests/test_epochs.py]: so a test for `flat` would be good I guess if it applies to both ...

File: mne/epochs.py [hunk 6]
```diff
@@ -444,49 +444,54 @@ def apply_baseline(self, baseline, verbose=None):
...
-                                                    new=new, old=old, op='<'))
+            if len(idx[key]) == 0 and (key in reject or key in flat):
+                # This is where we could eventually add e.g.
+                # self.allow_missing_reject_keys check to allow users to
+                # provide keys that don't exist in data
+                raise ValueError("No %s channel found. Cannot reject based on "
+                                 "%s." % (key.upper(), key.upper()))
+
+        # now check to see if our rejection and flat are getting more
+        # restrictive
```
Comment (jasmainak) [inline mne/epochs.py]: so the variable `key` from the loop was never used inside the loop? is that why you moved the following lines outside the loop?

Comment (Eric89GXL) [inline mne/epochs.py]: yeah we were repeating the same check a bunch of times for no reason

**Comment by Eric89GXL** [comment]
Thanks for the reviews, comments addressed.

**Comment by jasmainak** [comment]
you need to make travis happy

**Comment by Eric89GXL** [comment]
Sorry had a bad commit (and killed CIs), push forced the correct one.

**Comment by jasmainak** [comment]
LGTM. +1 for merge if Travis is happy

**Comment by agramfort** [comment]
thx @Eric89GXL !
